### PR TITLE
Gate ubuntu-22.04 in CI and stop the assign harness hiding failures (Fixes #2688)

### LINKED ISSUE #2688: Nightly release fails on ubuntu-22.04: assign harness breaks on an untested runner platform
## Summary

The nightly Release workflow failed on run 30139165201 with 12 failed test files in the script harness (Test Files 12 failed | 96 passed | 5 skipped). All failures are in the "Run Preflight Checks" step, which chains `npm run test:ci` -> `npm run test:scripts`.

The underlying problem is a **platform coverage gap**: the nightly is the only workflow that runs on `ubuntu-22.04`, and it is not a required PR check. So a whole runner platform is exercised only where nobody is watching, and regressions surface at 01:46 instead of on the PR that caused them.

## Evidence

Failing run: https://github.com/vybestack/llxprt-code/actions/runs/30139165201/job/89628961760

Runner image: `ubuntu-22.04`, version 20260720.234.2

Platform matrix today:

| Workflow | Runner(s) |
| --- | --- |
| ci.yml (PR gate) | ubuntu-latest, macos-latest |
| release.yml (nightly) | **ubuntu-22.04** |

`grep -c ubuntu-22.04 .github/workflows/ci.yml` returns 0. `ubuntu-22.04` appears in `release.yml` and nowhere else, so it has zero PR coverage.

## Failure breakdown

12 failed files. One is already fixed on main:

- `ocr-review-workflow-behaviors.test.js` — "expected '1.7.16' to be '1.7.9'". This is the stale OCR_VERSION assertion from #2679, fixed in #2682 (merged as 6bc801020). It is listed here only for completeness.

The other 11 are all `assign-*` files and **all share one function, `runCleanup`**, from `scripts/tests/assign-helpers.js`:

    assign-remediation.test.js       3 failed
    assign-remediation2.test.js     11 failed
    assign-remediation3.test.js      3 failed
    assign-remediation4.test.js      3 failed
    assign-remediation5.test.js      2 failed
    assign-remediation6.test.js      2 failed
    assign-remediation7.test.js      4 failed
    assign-remediation8.test.js      2 failed
    assign-remediation10.test.js     3 failed
    assign-remediation11.test.js     6 failed
    assign-workflow-behaviors.test.js 7 failed

Every one shows the same signature: the cleanup script exits nonzero having applied **zero mutations**. Representative assertions:

    AssertionError: expected 1 to be +0        (result.status === 0 failed; script exited 1)
    AssertionError: expected [ 'stale-user' ] to include 'new-user'
    AssertionError: expected [] to have a length of 1 but got +0

These fail **fast** (assign-remediation10 completed in 1440ms), so they are NOT timeouts. This is not related to the testTimeout change in #2682.

## Ruled out (with evidence, not assumption)

- **Not my #2682 change / not a merge regression.** The nightly ran on 3a12f02e5, which `git merge-base --is-ancestor 3a12f02e5 6bc801020` confirms predates the #2682 merge.
- **Not clock drift.** Dates are fully deterministic: ASSIGN_NOW is pinned to 2025-07-23T00:00:00Z in assign-helpers.js and all seeded timestamps are literals.
- **Not disk/inode/fd exhaustion.** No ENOSPC, "No space left", EMFILE, or "too many open files" anywhere in the log.
- **Not a Python version issue.** fake-gh.py contains no 3.11+ only constructs (no datetime.UTC, match/case, tomllib, Self, StrEnum).
- **Not a macOS-only or load-related flake.** The full set of failing files passes locally on macOS: 124/124 across all 6 primary files run concurrently, and assign-remediation2 passes 26/26 in isolation.
- **The "gh: command failed" line in the log is an intentional negative test** in aggregate-historical.test.js, not a real failure.

## Not yet isolated

The exact `ubuntu-22.04` trigger is not pinned down. Docker has no running daemon on the investigating machine, so the image could not be reproduced locally. The leading candidate is a **toolchain version difference** — most likely jq (1.6 on ubuntu-22.04 vs 1.7.x on ubuntu-latest and locally). `unassign-stale-issues.sh` has 34 jq invocations including `@uri` and several `--argjson` filters.

## Why this was hard to diagnose (and the second bug)

`runCleanup` in `scripts/tests/assign-helpers.js` returns `stderr: ''` on the success path, and on the failure path the tests assert only on status and state. The cleanup script prints a real diagnostic to stderr, for example:

    printf '‼ Candidate discovery failed: %s\n' "${_diag}" >&2

but that never reaches the test output. Twelve failing files produced **zero actionable cause**. That is a diagnosability defect in its own right, independent of whatever the 22.04 trigger turns out to be.

## Proposed fix

1. **Close the platform gap.** Add `ubuntu-22.04` to the ci.yml test matrix so the release platform is gated on PRs, rather than aligning the nightly to ubuntu-latest (which would leave the actual release runner untested).
2. **Make the harness self-diagnosing.** Surface the cleanup/assign script stderr in failure output so the next occurrence reports its own root cause instead of requiring log archaeology.

Item 2 is expected to reveal the precise 22.04 trigger, which can then be fixed directly.

## Context

Discovered while investigating #2668. Same class of defect: #2668 was a macOS-only test gate that manufactured platform-specific-looking failures; this is an ubuntu-22.04 platform with no PR coverage at all. Both hide real bugs behind a coverage blind spot.

### PR DESCRIPTION
**Pull request by acoliver**
## Summary

Closes #2688.

The nightly release job runs on `ubuntu-22.04`. No other workflow used that platform, so it had **zero PR coverage**. When the assign harness broke there, twelve test files failed with nothing but bare assertions like "expected 1 to be +0", discovered at 01:46 against a commit that had already merged.

This PR closes the coverage gap and fixes the reason the failure was expensive to diagnose.

## The coverage gap

| Workflow | Runner(s) |
| --- | --- |
| ci.yml (PR gate) | ubuntu-latest, macos-latest |
| release.yml (nightly) | **ubuntu-22.04** |

`grep -c ubuntu-22.04 .github/workflows/ci.yml` returned 0 before this change.

Added `ubuntu-22.04` to the ci.yml test matrix (3 OS x 2 secure-store modes = 6 legs).

I considered aligning the nightly to `ubuntu-latest` instead and **rejected it**: that would leave the actual release runner untested, trading one blind spot for another. Verified that artifact names are already keyed by `matrix.os`, so the added leg does not collide, and that `post_coverage_comment` consumes only the `ubuntu-latest` artifact and is unaffected.

## The diagnosability defect

`execFileSync` only returns stdout, so **stderr from a successful run was unreachable**, and all three call sites hardcoded `stderr: ''` on the success path. The bash scripts report their real root cause on stderr:

    printf '‼ Candidate discovery failed: %s\n' "${_diag}" >&2

None of it reached the tests. Twelve failing files produced zero actionable cause.

Switched to `spawnSync` behind a single shared `runAutomationScript` helper, which also removes three near-identical try/catch blocks (net -48 lines of duplication). A signal-terminated child reports `status === null`; that is treated as failure rather than coerced to 0, so a killed script can never be reported as a clean pass.

Before and after, on the same injected failure:

    before: status=1  stderr=""
    after:  status=1  stderr="‼ Candidate discovery failed: {"message": "Server Error", ...}"

## Tests

Adds `scripts/tests/assign-harness-diagnostics.test.js` with three behavioral tests that execute the real bash scripts against the fake gh adapter and assert observable output.

Every test was **mutation-tested**, and the assertions were tightened twice in response to mutations that survived:

- Restoring the original `stderr: ''` fails all 3.
- Substituting generic text (`'some unrelated warning noise'`) for real stderr also fails all 3 -- so the tests pin the actual root-cause marker, not merely non-empty output.

Two earlier drafts were discarded during this process: a pair of tautological `typeof` assertions that caught nothing under mutation, and one test that asserted a success-with-warning path which **does not exist** (timeline failure exits 1, not 0). Both were replaced with assertions against verified real behavior rather than assumed behavior.

## Scope

This PR does **not** fix the underlying ubuntu-22.04 trigger, which is not yet isolated -- Docker has no running daemon on the investigating machine, so the image could not be reproduced locally. The leading candidate is a jq version difference (1.6 on 22.04 vs 1.7.x elsewhere; `unassign-stale-issues.sh` has 34 jq invocations including `@uri` and several `--argjson` filters).

That is deliberate. With this PR merged, the 22.04 leg runs on every PR **and** the harness reports the script's own error message, so the next occurrence should self-diagnose instead of requiring log archaeology. I expect this PR's own CI run to surface the trigger directly.

Note that one of the twelve nightly failures (`ocr-review-workflow-behaviors.test.js`, "expected '1.7.16' to be '1.7.9'") was the stale OCR_VERSION assertion from #2679, already fixed on main by #2682.

## Verification

- `npm run lint` -- pass
- `npm run typecheck` -- pass
- `npm run format` -- clean
- `npm run build` -- pass
- `npm run lint:eslint-guard` -- "ESLint policy guard passed"
- `npm run test:scripts` -- **109 files / 2869 tests passed, 0 failures**
- Smoke test (`bun scripts/start.ts --profile-load ollamakimi`) -- pass
- Open Code Review -- one medium finding on a too-weak assertion, addressed in 9c5f581a6; re-review clean ("Looks good to me")

No lint rules loosened, no suppression directives added, no complexity thresholds raised.

`actionlint` reports one SC2129 style finding in ci.yml; verified pre-existing on main (identical finding, line number shifted only by my added comment).

## Summary by CodeRabbit

- **Bug Fixes**
  - Enhanced failure diagnostics for assignment and cleanup automation, preserving error output and improving CI-only logging.
  - Improved handling for interrupted runs and clearer reporting of operation and root-cause details for specific issues.
- **Tests**
  - Added coverage to validate the assignment harness’s diagnosability under simulated infrastructure/API failures, including retry behavior and state handling.
  - Added workflow validation to ensure runner-image consistency (using `ubuntu-latest`) and to prevent orphaned runner images.
- **Chores**
  - Updated release workflow runner settings and added toolchain version logging to CI.

### CODE REVIEW

**Comment by acoliver** [comment]
**CI status: blocked by an unrelated pre-existing failure, now tracked as #2691.**

`Lint (Javascript)` fails on this PR, and because the `test` job depends on `lint`, the entire test matrix is skipped — including the new `ubuntu-22.04` legs this PR adds. So this run cannot yet demonstrate the intended effect.

The failure is not from this PR:

    error: Failed to resolve root dev dependency '@types/bun'
        at bun.lock:5:9
    InvalidPackageInfo: failed to parse lockfile: 'bun.lock'

Evidence it is pre-existing and repo-wide:

- **Reproduces on a pristine `main` worktree** with zero local modifications (`git worktree add /tmp/x main && cd /tmp/x && bun install`).
- This PR touches only `.github/workflows/ci.yml` and two files under `scripts/tests/` — no dependency manifest.
- Same failure on other current PR branches: `issue2666` (runs 30161918570, 30161750618, 30161381617) and `issue2683` (30161260726).

Filed as #2691 with a root cause and two verified repair options. Not bundled here to keep this PR scoped; full lockfile regeneration silently downgrades a number of packages, which deserves its own review.

`Lint (GitHub Actions)`, `Lint (YAML)`, and `Lint (Shell)` all pass, so the matrix change itself is syntactically valid. Once #2691 lands I will rebase and re-run so the ubuntu-22.04 legs actually execute.

File: scripts/tests/assign-harness-diagnostics.test.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        assignee: 'stale-user',
+        createdAt: daysAgo(20),
+      }),
+    ],
+  };
+  return createFakeRepo({ ...state, ...extraState });
+}
+
+describe('assign harness diagnosability (#2688)', () => {
+  it('surfaces the script\u2019s own root-cause message when cleanup fails', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/assign-harness-diagnostics.test.js]: Fixed in 258cc80e0. Confirmed no other file under scripts/tests uses \u2019 escapes, so this was inconsistent with the surrounding convention. Both test names now use literal apostrophes.

File: scripts/tests/assign-harness-diagnostics.test.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      makeAssignedEvent({
+        assignee: 'stale-user',
+        createdAt: daysAgo(20),
+      }),
+    ],
+  };
+  return createFakeRepo({ ...state, ...extraState });
+}
+
+describe('assign harness diagnosability (#2688)', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/assign-harness-diagnostics.test.js:60]: Good catch — this was a real gap. Fixed in 258cc80e0.

Added a test for a run that exits 0 after the retry loop recovers from a transient API error. That case emits a retry warning on stderr while still succeeding, which is precisely the shape of the original defect: stderr was unreachable on the success path, so a degraded-but-successful run looked completely silent.

Verified it closes the gap by mutation rather than assuming it. Erasing stderr only on success:

    stderr: status === 0 ? '' : (result.stderr ?? '')

fails the new test and no other (1 failed | 3 passed) — so the three pre-existing failure-path tests genuinely could not reach this, and the new one does.

One correction on the suggestion: asserting that a successful run returns stderr "even if empty" would be tautological, since an empty string is what the buggy version returned too. The test instead pins a success path that produces real stderr content, which is what actually distinguishes fixed from broken.

**Comment by acoliver** [comment]
**Update: the #2691 lockfile blocker is now fixed here in 388222610, so this PR is no longer blocked.**

I initially kept it separate to avoid conflating concerns, but leaving it out meant the `test` matrix stayed skipped and the `ubuntu-22.04` legs this PR exists to add never ran — so the PR could not demonstrate its own effect. Fixing it here rather than leaving CI red.

The repair is deliberately minimal: it inserts only the two missing resolution entries (`@types/bun` and its dependency `bun-types`), using integrity hashes taken from the registry.

I rejected full regeneration (`rm bun.lock && bun install`). It fixes the parse error too, but rewrites 564 lines, reverts `configVersion` 1 -> 0, and **silently downgrades** packages including `@ai-sdk/openai` 2.0.110 -> 2.0.109 and `@azure/msal-browser` 5.16.0 -> 5.15.0. Version changes like that should be a deliberate decision, not a side effect of a lockfile repair.

Verified mechanically rather than by eye — comparing every resolved version against HEAD:

    resolved-version CHANGES: 0
    added keys: ['@types/bun', 'bun-types', ...workspace renormalization...]

Zero resolved versions changed. The remaining additions/removals are Bun's own workspace `file:packages/*` renormalization, which is documented in dev-docs/bun.md.

Post-fix verification: `bun install` clean (parse error gone), `npm run typecheck` pass, `npm run build` pass.
